pkg/kubelet/status/state/state_checkpoint_test.go

```diff
@@ -0,0 +1,128 @@
+limitations under the License.
+*/
+
+package state
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
+)
+
+func Test_stateCheckpoint_storeState(t *testing.T) {
+	// create temp dir
+	testingDir, err := os.MkdirTemp("", "pod_resource_allocation_state_test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if err := os.RemoveAll(testingDir); err != nil {
+			t.Fatal(err)
+		}
+	}()
+	cache := NewStateMemory()
+	checkpointManager, err := checkpointmanager.NewCheckpointManager(testingDir)
+	require.NoError(t, err, "failed to create checkpoint manager")
+	checkpointName := "pod_state_checkpoint"
+	sc := &stateCheckpoint{
+		cache:             cache,
+		checkpointManager: checkpointManager,
+		checkpointName:    checkpointName,
+	}
+	type args struct {
+		podResourceAllocation PodResourceAllocation
+	}
+
+	tests := []struct {
+		name string
+		args args
+	}{}
+	suffix := []string{"Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "n", "u", "m", "k", "M", "G", "T", "P", "E", ""}
+	factor := []string{"1", "0.1", "0.03", "10", "100", "512", "1000", "1024", "700", "10000"}
+	for _, fact := range factor {
+		for _, suf := range suffix {
+			if (suf == "E" || suf == "Ei") && (fact == "1000" || fact == "10000") {
+				// when fact is 1000 or 10000, suffix "E" or "Ei", the quantity value is overflow
+				// see detail https://github.com/kubernetes/apimachinery/blob/95b78024e3feada7739b40426690b4f287933fd8/pkg/api/resource/quantity.go#L301
+				continue
+			}
+			tests = append(tests, struct {
+				name string
+				args args
+			}{
+				name: fmt.Sprintf("resource - %s%s", fact, suf),
+				args: args{
+					podResourceAllocation: PodResourceAllocation{
+						"pod1": {
+							"container1": {
+								v1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+								v1.ResourceMemory: resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+							},
+						},
+					},
+				},
+			})
+		}
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err = sc.cache.ClearState()
+			require.NoError(t, err, "failed to clear state")
+
+			defer func() {
+				err = checkpointManager.RemoveCheckpoint(checkpointName)
+				require.NoError(t, err, "failed to remove checkpoint")
+			}()
+
+			err = sc.cache.SetPodResourceAllocation(tt.args.podResourceAllocation)
+			require.NoError(t, err, "failed to set pod resource allocation")
+
+			err = sc.storeState()
+			require.NoError(t, err, "failed to store state")
+
+			// deep copy cache
+			originCache := NewStateMemory()
+			podAllocation := sc.cache.GetPodResourceAllocation()
+			err = originCache.SetPodResourceAllocation(podAllocation)
+			require.NoError(t, err, "failed to set pod resource allocation")
+
+			err = sc.cache.ClearState()
+			require.NoError(t, err, "failed to clear state")
+
+			err = sc.restoreState()
```

Comment:
Would it be better to test `storeState` and `restoreState` separately, e.g. [as it's done for DRA](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/state/checkpointer_test.go)

Reply:
This would make the tests easier to understand, wouldn't require deep copying and would help to avoid hidden issues that might be compensating each other when `storeState` and `restoreState` are called in the same test.

Reply:
Actually, the tests are designed for verifying the whole flow, to make sure the content written to the file is the samw with the content read from the file. If we split them into separate test, the error might be hidden, just like the UT in description of this PR. https://github.com/kubernetes/kubernetes/pull/126620#issue-2459128710

Reply:
Well, generally unit tests are designed to test units. If you want to test whole worklow please provide e2e tests as well.

Reply:
@bart0sh What do you think?

Reply:
It's up to you. I don't feel strongly about this.